File: exercise series (2019–2020)/week_10 [JavaScript]/Chaocipher.py
class Disk:
    def __init__(self, alphabet: str):
        from string import ascii_uppercase
        self.alphabet = alphabet.upper()
        assert ''.join(sorted(self.alphabet)) == 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', "invalid alphabet"

    def __str__(self):
        return "".join(self.alphabet)

    def locate(self, letter: str):
        return self.alphabet.index(letter.upper()) + 1

    def letter(self, index: int):
        return self.alphabet[index - 1]

    def rotate(self, count: int):
        count = ((count % 26) + 26) % 26
        self.alphabet = self.alphabet[count:] + self.alphabet[:count]
        return self

    def roll(self, start: int, stop: int):
        self.alphabet = (self.alphabet[:start - 1]
                         + self.alphabet[start:stop]
                         + self.alphabet[start - 1]
                         + self.alphabet[stop:])
        return self


class Chaocipher:
    def __init__(self, left_alphabet: str, right_alphabet):
        self.left = Disk(left_alphabet)
        self.right = Disk(right_alphabet)

    def __str__(self):
        return f"""            +            *
 LEFT (ct): {self.left}
RIGHT (pt): {self.right}
            --------------------------"
  Position: 12345678911111111112222222"
                     01234567890123456"""

    def encode(self, plaintext: str):
        encoded = ""
        for letter in plaintext:
            index = self.right.locate(letter)
            encoded += self.left.letter(index)
            self.permute(index)
        return encoded

    def decode(self, encoded: str):
        decoded = ""
        for letter in encoded:
            index = self.left.locate(letter)
            decoded += self.right.letter(index)
            self.permute(index)
        return decoded

    def permute(self, index):
        self.left.rotate(index - 1).roll(2, 14)
        self.right.rotate(index).roll(3, 14)

File: exercise series (2019–2020)/week_10 [JavaScript]/Chaocipher.js
class Disk {
    constructor(alphabet) {
        this.alphabet = alphabet.toUpperCase();
        if ([...this.alphabet].sort().join("") !== "ABCDEFGHIJKLMNOPQRSTUVWXYZ") {
            throw {name: "AssertionError", message: "invalid alphabet"}
        }
    }

    toString() {
        return this.alphabet;
    }

    locate(letter) {
        return this.alphabet.indexOf(letter.toUpperCase()) + 1;
    }

    letter(index) {
        return this.alphabet[index - 1];
    }

    rotate(count) {
        count = ((count % 26) + 26) % 26;
        this.alphabet = this.alphabet.slice(count) + this.alphabet.slice(0, count);
        return this;
    }

    roll(start, stop) {
        this.alphabet = this.alphabet.slice(0, start - 1)
            + this.alphabet.slice(start, stop)
            + this.alphabet[start - 1]
            + this.alphabet.slice(stop);
        return this;
    }
}

class Chaocipher {
    constructor(leftAlphabet, rightAlphabet) {
        this.left = new Disk(leftAlphabet);
        this.right = new Disk(rightAlphabet);
    }

    toString() {
        return `            +            *`
            + `\n LEFT (ct): ${this.left.toString()}`
            + `\nRIGHT (pt): ${this.right.toString()}`
            + `\n            --------------------------`
            + `\n  Position: 12345678911111111112222222`
            + `\n                     01234567890123456`
    }

    encode(plaintext) {
        let encoded = "";
        for (const letter of plaintext) {
            const index = this.right.locate(letter);
            encoded += this.left.letter(index);
            this.permute(index);
        }
        return encoded;
    }

    decode(encoded) {
        let decoded = "";
        for (const letter of encoded) {
            const index = this.left.locate(letter);
            decoded += this.right.letter(index);
            this.permute(index);
        }
        return decoded;
    }

    permute(index) {
        this.left.rotate(index - 1).roll(2, 14);
        this.right.rotate(index).roll(3, 14)
    }
}